I say NO to magic number in the code. Be smart, use NUM_PIXELS!

dglaude opened the pull request:
Let's show the good example and try to use the [documented constant](http://docs.pimoroni.com/blinkt/#constants) !

Now Pimoroni can release their 16 LED Blinkt!

dglaude committed:
I say NO to magic number in the code. Be smart, use NUM_PIXELS!

Gadgetoid commented on examples/cpu_load.py:
```diff
@@ -9,13 +9,13 @@
     exit("This script requires the psutil module\nInstall with: sudo pip install psutil")
 
-from blinkt import set_clear_on_exit, set_brightness, set_pixel, show
+from blinkt import set_clear_on_exit, set_brightness, set_pixel, show, NUM_PIXELS
```
It's only tangentially related to your PR, but I think to better support this change, all instances of explicit imports should be replaced with:

```
import blinkt
```

And then `blinkt.set_clear_on_exit`, `blinkt.set_brightness`, `blinkt.set_pixel`, `blinkt.NUM_PIXELS`, etc.

Gadgetoid commented on examples/candle.py:
```diff
@@ -21,10 +21,10 @@
+        limit = NUM_PIXELS
     for pixel in range(limit):
-        hue = start + (((end - start) / 8.0) * pixel) 
+        hue = start + (((end - start) / (NUM_PIXELS*1.0)) * pixel) 
```
I think `float(NUM_PIXELS)` would be preferable to `(NUM_PIXELS*1.0)` here (technically correct, I guess?) since it's an implicit coercion to float, rather than relying on the implicit side-effect of multiplying an integer by a float.

Gadgetoid commented on examples/gradient_graph.py:
```diff
@@ -13,9 +13,9 @@
-        hue = ((hue_start + ((x/8.0) * hue_range)) % 360) / 360.0
+    v *= NUM_PIXELS
+    for x in range(NUM_PIXELS):
+        hue = ((hue_start + ((x/(NUM_PIXELS*1.0)) * hue_range)) % 360) / 360.0
```
Again, `float(NUM_PIXELS)` is preferable.

Gadgetoid commented on examples/larson_hue.py:
```diff
@@ -28,11 +28,11 @@
-        val = 7 - (abs(offset - x) * FALLOFF)
-        val /= 7.0 # Convert to 0.0 to 1.0
+        val = (NUM_PIXELS-1) - (abs(offset - x) * FALLOFF)
+        val /= ((NUM_PIXELS-1)*1.0) # Convert to 0.0 to 1.0
```
While we're tidying things up, this might be made clearer by declaring a `max_val = NUM_PIXELS-1` outside of the loop and then:

```python
val = max_val - (abs(offset - x) * FALLOFF)
val /= float(max_val)
```

Gadgetoid reviewed (commented):
I think in general this is the right approach, although there are no hard-and-fast rules when it comes to writing clear code.

I've made some suggestions for changes (namely the casting to float) and some other changes that may compliment this set nicely.

dglaude commented:
Just to indicate that before those change to Blinkt!, I made a Pull request
to Rainbow HAT that ported some of those example to 7 APA102 rather than 8.

I'll try to rework both in a compatible manner where almost the same code
will work on various number of multi-color LED.

I am thinking:
* 7 for Rainbow HAT
* 8 Blinkt
* 16 for a single Mote stick using pHAT mote
* 16 for a single Mote stick using Mote

I even feel like RasPiO InsPiring could benefit from this... sorry this is
not one of your product? ;-)

I am a bit frustrated to see very few example on new board and duplication
of effort in making example alike but not exactly the same.

I see 3 families of products:
* 1D colour LEDs (what we talk about)
* 2D colour LEDs (the job was already done for Unicorn HAT and Unicorn pHAT)
* 2D mono colour LEDs (with Scroll pHAT and Scroll pHAT HD)

Multiple MOTE stick could be either a very long 1D colour LED strip or a 2D
matrix, like a long Unicorn pHAT.


2017-03-22 12:37 GMT+01:00 Philip Howard <notifications@github.com>:

> *@Gadgetoid* commented on this pull request.
>
> I think in general this is the right approach, although there are no
> hard-and-fast rules when it comes to writing clear code.
>
> I've made some suggestions for changes (namely the casting to float) and
> some other changes that may compliment this set nicely.
> ------------------------------
>
> In examples/cpu_load.py
> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107389571>:
>
> > @@ -9,13 +9,13 @@
>  except ImportError:
>      exit("This script requires the psutil module\nInstall with: sudo pip install psutil")
>
> -from blinkt import set_clear_on_exit, set_brightness, set_pixel, show
> +from blinkt import set_clear_on_exit, set_brightness, set_pixel, show, NUM_PIXELS
>
> It's only tangentially related to your PR, but I think to better support
> this change, all instances of explicit imports should be replaced with:
>
> import blinkt
>
> And then blinkt.set_clear_on_exit, blinkt.set_brightness, blinkt.set_pixel,
> blinkt.NUM_PIXELS, etc.
> ------------------------------
>
> In examples/candle.py
> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107390051>:
>
> >      for pixel in range(limit):
> -        hue = start + (((end - start) / 8.0) * pixel)
> +        hue = start + (((end - start) / (NUM_PIXELS*1.0)) * pixel)
>
> I think float(NUM_PIXELS) would be preferable to (NUM_PIXELS*1.0) here
> (technically correct, I guess?) since it's an implicit coercion to float,
> rather than relying on the implicit side-effect of multiplying an integer
> by a float.
> ------------------------------
>
> In examples/gradient_graph.py
> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107390215>:
>
> > @@ -13,9 +13,9 @@
>  max_brightness = 0.2
>
>  def show_graph(v, r, g, b):
> -    v *= 8
> -    for x in range(8):
> -        hue = ((hue_start + ((x/8.0) * hue_range)) % 360) / 360.0
> +    v *= NUM_PIXELS
> +    for x in range(NUM_PIXELS):
> +        hue = ((hue_start + ((x/(NUM_PIXELS*1.0)) * hue_range)) % 360) / 360.0
>
> Again, float(NUM_PIXELS) is preferable.
> ------------------------------
>
> In examples/larson_hue.py
> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107390854>:
>
> >          sat = 1.0
>
> -        val = 7 - (abs(offset - x) * FALLOFF)
> -        val /= 7.0 # Convert to 0.0 to 1.0
> +        val = (NUM_PIXELS-1) - (abs(offset - x) * FALLOFF)
> +        val /= ((NUM_PIXELS-1)*1.0) # Convert to 0.0 to 1.0
>
> While we're tidying things up, this might be made clearer by declaring a max_val
> = NUM_PIXELS-1 outside of the loop and then:
>
> val = max_val - (abs(offset - x) * FALLOFF)
> val /= float(max_val)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pimoroni/blinkt/pull/47#pullrequestreview-28356444>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASiRnJ0BcGvgVPHJrJKIfClNCUe6Hammks5roQfegaJpZM4MjJiu>
> .
>

dglaude commented:
I had an idea for portability:

import blinkt as apa

And then

apa.set_clear_on_exit()

This would make the code compatible with Rainbow HAT by just changing

import blinkt as apa

by

import rainbow as apa

What do you think about that?


On Mar 22, 2017 19:09, "David Glaude" <david.glaude@gmail.com> wrote:

> Just to indicate that before those change to Blinkt!, I made a Pull
> request to Rainbow HAT that ported some of those example to 7 APA102 rather
> than 8.
>
> I'll try to rework both in a compatible manner where almost the same code
> will work on various number of multi-color LED.
>
> I am thinking:
> * 7 for Rainbow HAT
> * 8 Blinkt
> * 16 for a single Mote stick using pHAT mote
> * 16 for a single Mote stick using Mote
>
> I even feel like RasPiO InsPiring could benefit from this... sorry this is
> not one of your product? ;-)
>
> I am a bit frustrated to see very few example on new board and duplication
> of effort in making example alike but not exactly the same.
>
> I see 3 families of products:
> * 1D colour LEDs (what we talk about)
> * 2D colour LEDs (the job was already done for Unicorn HAT and Unicorn
> pHAT)
> * 2D mono colour LEDs (with Scroll pHAT and Scroll pHAT HD)
>
> Multiple MOTE stick could be either a very long 1D colour LED strip or a
> 2D matrix, like a long Unicorn pHAT.
>
>
> 2017-03-22 12:37 GMT+01:00 Philip Howard <notifications@github.com>:
>
>> *@Gadgetoid* commented on this pull request.
>>
>> I think in general this is the right approach, although there are no
>> hard-and-fast rules when it comes to writing clear code.
>>
>> I've made some suggestions for changes (namely the casting to float) and
>> some other changes that may compliment this set nicely.
>> ------------------------------
>>
>> In examples/cpu_load.py
>> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107389571>:
>>
>> > @@ -9,13 +9,13 @@
>>  except ImportError:
>>      exit("This script requires the psutil module\nInstall with: sudo pip install psutil")
>>
>> -from blinkt import set_clear_on_exit, set_brightness, set_pixel, show
>> +from blinkt import set_clear_on_exit, set_brightness, set_pixel, show, NUM_PIXELS
>>
>> It's only tangentially related to your PR, but I think to better support
>> this change, all instances of explicit imports should be replaced with:
>>
>> import blinkt
>>
>> And then blinkt.set_clear_on_exit, blinkt.set_brightness,
>> blinkt.set_pixel, blinkt.NUM_PIXELS, etc.
>> ------------------------------
>>
>> In examples/candle.py
>> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107390051>:
>>
>> >      for pixel in range(limit):
>> -        hue = start + (((end - start) / 8.0) * pixel)
>> +        hue = start + (((end - start) / (NUM_PIXELS*1.0)) * pixel)
>>
>> I think float(NUM_PIXELS) would be preferable to (NUM_PIXELS*1.0) here
>> (technically correct, I guess?) since it's an implicit coercion to float,
>> rather than relying on the implicit side-effect of multiplying an integer
>> by a float.
>> ------------------------------
>>
>> In examples/gradient_graph.py
>> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107390215>:
>>
>> > @@ -13,9 +13,9 @@
>>  max_brightness = 0.2
>>
>>  def show_graph(v, r, g, b):
>> -    v *= 8
>> -    for x in range(8):
>> -        hue = ((hue_start + ((x/8.0) * hue_range)) % 360) / 360.0
>> +    v *= NUM_PIXELS
>> +    for x in range(NUM_PIXELS):
>> +        hue = ((hue_start + ((x/(NUM_PIXELS*1.0)) * hue_range)) % 360) / 360.0
>>
>> Again, float(NUM_PIXELS) is preferable.
>> ------------------------------
>>
>> In examples/larson_hue.py
>> <https://github.com/pimoroni/blinkt/pull/47#discussion_r107390854>:
>>
>> >          sat = 1.0
>>
>> -        val = 7 - (abs(offset - x) * FALLOFF)
>> -        val /= 7.0 # Convert to 0.0 to 1.0
>> +        val = (NUM_PIXELS-1) - (abs(offset - x) * FALLOFF)
>> +        val /= ((NUM_PIXELS-1)*1.0) # Convert to 0.0 to 1.0
>>
>> While we're tidying things up, this might be made clearer by declaring a max_val
>> = NUM_PIXELS-1 outside of the loop and then:
>>
>> val = max_val - (abs(offset - x) * FALLOFF)
>> val /= float(max_val)
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/pimoroni/blinkt/pull/47#pullrequestreview-28356444>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ASiRnJ0BcGvgVPHJrJKIfClNCUe6Hammks5roQfegaJpZM4MjJiu>
>> .
>>
>
>

dglaude committed:
Use flo(x) rathther than x*1.0
Use import blinkt and blinkt.xxx

Tentative on larson_hue.py to do
import blinkt as apa

Gadgetoid commented:
> import blinkt as apa

I think it's best to leave portability as an exercise to the user, rather than importing under an alias from the get go. It's already cringy enough seeing people commonly using `import unicornhat as uh` and `import numpy as np`. These aren't necessarily _wrong_, but they make for less explicit code, which I find ugly.

If you wanted to port Blinkt! examples to Rainbow HAT you could, in theory, just replace `import blinkt` with `import rainbowhat.rainbow as blinkt`.

I've been giving a not insignificant amount of thought to the idea of a unified drawing library and set of API calls for handling displays. It would be nice, but has a few practical problems:

* A new API across all libraries would give us *two* libraries to maintain for every product, there's only really one of me. Despite liberal help from @sandyjmacdonald and @RogueM, that's still a heck of a bottleneck.
* Such a unified library or API would only actually be useful to a tiny portion of our customers- not many people are moving code between products
* Most of the time products have different practical uses or intents, so unified code is of limited use and would tend to be over-complicated*


*EG: Trying to make an example fit both Blinkt! and Rainbow HAT is nice on the face of it, but I'd *much* rather see creative Rainbow HAT specific examples that make use of the buttons, temperature sensor and display too.

All this said, if we can iron out some of the differences between Mote pHAT, Mote, Blinkt! and RainbowHAT's Rainbow just in the name of sanity, that's a good thing.

AndiCui commented:
Maybe try to catch the exception of undefined NUM_PIXELS and set it to 8 to provide source compatibility (or at least print a helpful message).
Should the name of this variable begin with "_" to denote "no touchy" (See PEP 8)

dglaude committed:
Use import "blinkt" rather than "from blinkt import xxx"
List example that do not work with python3.

dglaude committed:
More usage of import "blinkt".
Remove ^M from readme.

dglaude commented:
@AndiCui Sorry, I don't get it.
NUM_PIXELS is defined in the library, so there is no reason for it not to be present.
NUM_PIXELS is a constant, that's why it is all CAP.
It is to be used, not something internal so why should it start with "_"?

dglaude commented:
I believe all of the example now use "import blinkt".
But maybe some example still don't use NUM_PIXELS.

I also detected that some example do not work with python3 (my prefered version):
cheerlights.py
cpu_temp.py
resistor_clock.py
twitter_monitor.py

I was unable to test some of the example like mqtt or the one requiring other pHAT.
I believe some exemple that require internet might be broken because the service does not exist or work. But my change to the code should not be the reason for those problem.

AndiCui commented:
Sorry, @dglaude, I confused myself again. Plz ignore me.

dglaude commented:
@Gadgetoid 
Trying your suggestion to port Blinkt! example to Rainbow HAT and the following does not work:
`import rainbowhat.rainbow as blinkt
`
I get an error

> Traceback (most recent call last):
>   File "candle.py", line 15, in <module>
>     import rainbowhat.rainbow as blinkt
> ImportError: No module named 'rainbowhat.rainbow'
>

dglaude committed:
One more 8 remplaced by NUM_PIXELS

dglaude committed:
One more place where NUM_PIXELS should be used (inderectly).

dglaude committed:
Fix a problem when random color is (0,0,0) and make it generic to NUM_PIXELS

Gadgetoid commented:
Ah, `rainbow` is not a module but an reference to the module, so it would be: `from rainbowhat import rainbow as blinkt` or, counter-intuitively, `import rainbowhat.apa102 as blinkt`

dglaude commented:
#import blinkt
from rainbowhat import rainbow as blinkt

Make most blinkt example work great on the pixel from the Rainbow HAT.
At least those flexible that use NUM_PIXELS as many in my pull request.

2017-03-30 16:35 GMT+02:00 Philip Howard <notifications@github.com>:

> Ah, rainbow is not a module but an reference to the module, so it would
> be: from rainbowhat import rainbow as blinkt or, counter-intuitively, import
> rainbowhat.apa102 as blinkt
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pimoroni/blinkt/pull/47#issuecomment-290430355>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASiRnJcmgXw8w4vd5a8U6nIABL1G3bxvks5rq72pgaJpZM4MjJiu>
> .
>

Gadgetoid committed:
Merge branch 'master' into better_example

Gadgetoid commented:
Merged for a handful of reasons, but mostly because I agree with the use of `import blinkt` versus importing individual methods from `blinkt`, and would like our examples to teach good habits.

As nice as `from blinkt import show` is for brevity, if your code grows to any sort of scale or complexity, or introduces any code from other libraries, the `show` method could quickly be confused or overridden. The same applies to others. Explicit is better than implicit!

I think all Python 3 fixes are merged too, this gives us a good set of examples to ship with v0.1.1. Thank you to all involved.
